src/main/kotlin/io/github/mojira/arisa/ArisaMain.kt

```diff
@@ -97,32 +100,47 @@ fun initModules(config: Config, jiraClient: JiraClient): (Issue) -> Map<String,
             return@lambda emptyMap()
         }
         val attachmentModule = AttachmentModule(
-            runIfShadowAttachment(config[Arisa.shadow], "DeleteAttachment", ::deleteAttachment.partially1(jiraClient)),
+            run1IfShadow(config[Arisa.shadow], "DeleteAttachment", ::deleteAttachment.partially1(jiraClient)),
             config[Arisa.Modules.Attachment.extensionBlacklist].split(",")
         )
         val chkModule = CHKModule(
-            runIfShadow(
+            run0IfShadow(
                 config[Arisa.shadow],
                 "UpdateCHK",
                 ::updateCHK.partially1(issue).partially1(config[Arisa.CustomFields.chkField])
             )
         )
         val reopenAwaitingModule = ReopenAwaitingModule(
-            runIfShadow(config[Arisa.shadow], "ReopenIssue", ::reopenIssue.partially1(issue))
+            run0IfShadow(config[Arisa.shadow], "ReopenIssue", ::reopenIssue.partially1(issue))
         )
         val piracyModule = PiracyModule(
-            runIfShadow(config[Arisa.shadow], "ResolveAsInvalid", ::resolveAsInvalid.partially1(issue)),
-            runIfShadow(
+            run0IfShadow(config[Arisa.shadow], "ResolveAsInvalid", ::resolveAsInvalid.partially1(issue)),
+            run0IfShadow(
                 config[Arisa.shadow],
                 "AddComment",
                 ::addComment.partially1(issue).partially1(config[Arisa.Modules.Piracy.piracyMessage])
             ),
             config[Arisa.Modules.Piracy.piracySignatures].split(",")
         )
         val removeTriagedMeqsModule = RemoveTriagedMeqsModule(
-            runIfShadowComment(config[Arisa.shadow], "UpdateCommentBody", ::updateCommentBody.partially1(jiraClient)),
+            run2IfShadow(config[Arisa.shadow], "UpdateCommentBody", ::updateCommentBody.partially1(jiraClient)),
             config[Arisa.Modules.RemoveTriagedMeqs.meqsTags].split(",")
         )
+        val futureVersionModule = FutureVersionModule(
+            run1IfShadow(config[Arisa.shadow], "RemoveAffectedVersion", ::removeAffectedVersion.partially1(issue)),
+            run0IfShadow(config[Arisa.shadow],
+                "AddAffectedVersion"
+            ) {
+                val latestVersion = getLatestReleasedVersion(jiraClient.getProject(issue.project.key))
+                if (latestVersion.isLeft()) (latestVersion as Either.Left).a.left()
+                else addAffectedVersion(issue, (latestVersion as Either.Right).b)
+            },
```

Comment:
This will throw errors in MCTEST.
The reason for that is, MCTEST only contains unreleased versions. Because of this, a `NoSuchElementException` is thrown by `getLatestReleasedVersion`.
That shouldn't be a problem in other projects though.

Reply:
With errors you mean an actual exception or a error module response?

I was planning on having alerts for that, so if it might trigger regularly might be worth extracting it to another method and handling the exception (with a try catch or a Either.catch)

Reply:
https://github.com/mojira/arisa-kt/blob/82a4a8b6d7efb220d9e621970289a883e826aacf/src/main/kotlin/io/github/mojira/arisa/infrastructure/Jira.kt#L65-L69

The exception is wrapped by an Either.catch over here, resulting in a `FailedModuleResponse`:
```
7:05:08.446 [main] ERROR Arisa - [RESPONSE] [MCTEST-64] [FutureVersion] Failed
java.util.NoSuchElementException: List contains no element matching the predicate.
	at io.github.mojira.arisa.infrastructure.JiraKt$getLatestReleasedVersion$1$1.invokeSuspend(Jira.kt:94)
	at io.github.mojira.arisa.infrastructure.JiraKt$getLatestReleasedVersion$1$1.invoke(Jira.kt)
	at arrow.core.Either$Companion.catch(Either.kt:862)
	at arrow.core.Either$Companion.catch(Either.kt:858)
	at io.github.mojira.arisa.infrastructure.JiraKt$getLatestReleasedVersion$1.invokeSuspend(Jira.kt:66)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
.... (many more)
```

Reply:
Or well, I guess this is actually what's taking care of it in the end. That's the function for adding the latest version that's being passed to the module:
https://github.com/mojira/arisa-kt/blob/82a4a8b6d7efb220d9e621970289a883e826aacf/src/main/kotlin/io/github/mojira/arisa/ArisaMain.kt#L131-L137

Reply:
I would have an assert before that, assertAnyVersionIsReleased

In there you can check if project.versions.any { it.isReleased && !it.isArchived } and if thats false, return operationnotneeded

Reply:
Done.